rust-runtime/aws-smithy-runtime-api/src/shared.rs

```diff
@@ -0,0 +1,137 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+//! Conversion traits for converting an unshared type into a shared type.
+//!
+//! The standard [`From`](std::convert::From)/[`Into`](std::convert::Into) traits can't be
+//! used for this purpose due to the blanket implementation of `Into`.
+//!
+//! This implementation also adds a [`maybe_shared`] method and [`impl_shared_conversions`]
+//! macro to trivially avoid nesting shared types with other shared types.
+
+use std::any::{Any, TypeId};
+
+/// Like the `From` trait, but for converting to a shared type.
+pub trait FromUnshared<Unshared> {
+    /// Creates a shared type from an unshared type.
+    fn from_unshared(value: Unshared) -> Self;
+}
+
+/// Like the `Into` trait, but for (efficiently) converting into a shared type.
+///
+/// If the type is already a shared type, it won't be nested in another shared type.
+pub trait IntoShared<Shared> {
+    /// Creates a shared type from an unshared type.
+    fn into_shared(self) -> Shared;
+}
```

Comment:
I _think_ this should be an associated type? It doesn't make sense for one type to be convertible into multiple different shared types right?

Reply:
I basically made these traits equivalent to `From`/`Into`. I guess it makes sense to restrict it to one shared type, but typing `IntoShared<Shared = MySharedType>` everywhere kind of sucks. I wish we had trait aliases.